presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java

```diff
@@ -0,0 +1,107 @@
+    @Nullable
+    private final PagesReleasedListener childListener;
+    private final ConcurrentMap<Lifespan, AtomicLong> outstandingPageCountPerLifespan = new ConcurrentHashMap<>();
+    private final Set<Lifespan> noMorePagesForLifespan = ConcurrentHashMap.newKeySet();
+    private volatile Consumer<Lifespan> lifespanCompletionCallback;
+
+    public LifespanSerializedPageTracker(OutputBufferMemoryManager memoryManager)
+    {
+        this(memoryManager, Optional.empty());
+    }
+
+    public LifespanSerializedPageTracker(OutputBufferMemoryManager memoryManager, Optional<PagesReleasedListener> childListener)
+    {
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.childListener = requireNonNull(childListener, "childListener is null").orElse(null);
+    }
+
+    public boolean isLifespanCompletionCallbackRegistered()
+    {
+        return lifespanCompletionCallback != null;
+    }
+
+    public void registerLifespanCompletionCallback(Consumer<Lifespan> callback)
+    {
+        checkState(lifespanCompletionCallback == null, "lifespanCompletionCallback is already set");
+        this.lifespanCompletionCallback = requireNonNull(callback, "callback is null");
+    }
+
+    public void incrementLifespanPageCount(Lifespan lifespan, int pagesAdded)
+    {
+        AtomicLong counter = outstandingPageCountPerLifespan.get(lifespan);
+        if (counter == null) {
+            counter = outstandingPageCountPerLifespan.computeIfAbsent(lifespan, ignore -> new AtomicLong());
+        }
+        counter.addAndGet(pagesAdded);
+    }
+
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        requireNonNull(lifespan, "lifespan is null");
+        noMorePagesForLifespan.add(lifespan);
+    }
+
+    public boolean isNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        return noMorePagesForLifespan.contains(lifespan);
+    }
+
+    public boolean isFinishedForLifespan(Lifespan lifespan)
+    {
+        if (!noMorePagesForLifespan.contains(lifespan)) {
+            return false;
+        }
+
+        AtomicLong outstandingPageCount = outstandingPageCountPerLifespan.get(lifespan);
+        return outstandingPageCount == null || outstandingPageCount.get() == 0;
+    }
+
+    @Override
+    public void onPagesReleased(Lifespan lifespan, int releasedPageCount, long releasedSizeInBytes)
+    {
+        long outstandingPageCount = outstandingPageCountPerLifespan.get(lifespan).addAndGet(-releasedPageCount);
+        if (outstandingPageCount == 0 && noMorePagesForLifespan.contains(lifespan)) {
+            Consumer<Lifespan> lifespanCompletionCallback = this.lifespanCompletionCallback;
+            checkState(lifespanCompletionCallback != null, "lifespanCompletionCallback is not null");
+            lifespanCompletionCallback.accept(lifespan);
+        }
+        memoryManager.updateMemoryUsage(-releasedSizeInBytes);
+        if (childListener != null) {
```

Comment:
I didn't see this `childListener` related logic in https://github.com/trinodb/trino/pull/6097/ ? And looks like it's only used in `BroadcastOutputBuffer` as a safety check: 

```java
checkState(totalBufferedPages.addAndGet(-releasedPageCount) >= 0);
```

Reply:
That's correct. Since trino doesn't perform that sanity check there wasn't a need to add the same sub-listener logic. I'm not sure you'd want to add another atomic counter to the other buffers as they all have a more straightforward way of counting currently buffered pages (eg: `bufferedPages.size()`)